Caching

tylerstillwater opened the pull request:
MemCache integration.

@nelsam let's use this for discussion so we can get this completed. Please take a look at what I have so far and let me know of any problems you see.

As it stands now, everything should be in place to store and retrieve from cache. Next up is eviction.

tylerb committed:
Initial caching implementation

tylerb committed:
Custom encoder for caching

tylerb committed:
Fix aggressive regex replace

nelsam commented on plans/query_plans.go:
```diff
@@ -708,11 +727,32 @@ func (plan *QueryPlan) Select() ([]interface{}, error) {
 	}
-	return plan.executor.Select(target, query, plan.args...)
+	res, err := plan.executor.Select(target, query, plan.args...)
+
```
We need to check `err` here - if it's non-nil, we don't want to store the cache entry.

nelsam commented on plans/query_plans.go:
```diff
@@ -708,11 +727,32 @@ func (plan *QueryPlan) Select() ([]interface{}, error) {
+		cacheKey := fmt.Sprintf("%s: %v", query, plan.args)
+		data, err := getCacheData(cacheKey, plan.memCache)
+		if err == nil { // fail silently - graceful fallback
+			return data, nil
```
`data` needs to be unmarshaled to a slice of the same type as `plan.target`, here.  Otherwise, the query is going to return a different type depending on whether or not the data is in cache.

tylerb committed:
Unmarshal to slice of correct type

nelsam commented on plans/caching.go:
```diff
@@ -0,0 +1,116 @@
+		return nil, err
+	}
+
+	typed := reflect.Zero(reflect.SliceOf(targetType.Type())).Interface()
```
You don't actually need `reflect.Zero`, here.  `reflect.SliceOf(targetType.Type())` gives you the same value as `reflect.Zero(reflect.SliceOf(targetType.Type()))`.

tylerstillwater replied:
Right, but I cannot call `.Interface()` on it. `.SliceOf` gives `reflect.Type` where `.Zero` around it gives `reflect.Value`.

nelsam replied:
Ah ... I didn't even realize that, for some reason.  That seems odd, but meh.

nelsam commented on plans/caching.go:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	tv := reflect.ValueOf(typed)
+	data := reflect.MakeSlice(reflect.SliceOf(reflect.TypeOf([]interface{}{})), tv.Len(), tv.Cap()).Interface().([]interface{})
```
I think this should be `reflect.MakeSlice(reflect.TypeOf([]interface{}{}), tv.Len(), tv.Cap()).Interface()` - the `reflect.SliceOf` call seems unnecessary.

tylerb committed:
Add eviction

tylerb committed:
Working on proper cache decoding

tylerb committed:
Use map for storing cacheKeys. Overwrites existing.

tylerb committed:
Change decodeFromMemcache

nelsam commented on plans/caching.go:
```diff
@@ -0,0 +1,158 @@
+		elem = reflect.Zero(target.Type())
+	} else {
+		elem = reflect.New(target.Type().Elem())
+	}
```
Actually, the conditions here should be reversed - if it's a pointer, you want to use `reflect.New`; if it's not a pointer, you want to use `reflect.Zero`.  I probably should have caught that earlier.

tylerb committed:
Fix inverted logic

tylerb committed:
Optimize encoder

tylerb committed:
Invalidate types bound to target type

tylerb committed:
Ensure cache object is not nil

tylerb committed:
Caching now somewhat working

tylerb committed:
cache expire in 1 week

tylerb committed:
Fix cache key and table

tylerb committed:
improve debug output

nelsam committed:
Merge branch 'master' into caching

nelsam committed:
Changed up the caching logic

I moved all of the caching code around.  I think it's a bit cleaner, now.

I've also updated most of the cache set calls to run in a separate thread, so as to not prevent the query from returning.

nelsam committed:
Added sugar for cache settings on structs

tylerb committed:
Fix tests

tylerb committed:
Fix missing conditions for when to cache

tylerb committed:
Fix race condition

tylerb committed:
Use table name as map key

tylerb committed:
Fix data race in plan.args

tylerb committed:
Fix more data races in plan.args

tylerb committed:
remove prints

tylerb committed:
Fix race condition in plan.args by copying array

tylerb committed:
Add JoinOps() method

nelsam committed:
Revert "Use table name as map key"

This reverts commit 3ba99230a32bc88406df6e9b01080e24c70805e9.

nelsam committed:
Use anonymous TableMap pointer instead of value

We were running into issues with the TableMap value having a different address than the TableMap in our DbMap, making everything seem uncacheable.  The original fix was to use table names instead, but I think this is a cleaner fix.

tylerstillwater commented:
@nelsam all tests are passing in the api.
